### PR TITLE
Add model optimizations for codegen and gpt_bigcode

### PR DESCRIPTION
**Pull request by PhillipHoward**
# What does this PR do?

Optimizes codegen and gpt_bigcode to support static shapes with kv caching

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by sywangyi** [comment]
@regisss @fxmarty should you help review the PR, we enable codegen for habana both for inference and DDP finetune.

File: optimum/habana/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py [hunk 1]
```diff
@@ -0,0 +1,463 @@
...
+    self,
+    hidden_states: torch.Tensor,
+    layer_past: Optional[torch.Tensor] = None,
+    attention_mask: Optional[torch.Tensor] = None,
+    head_mask: Optional[torch.Tensor] = None,
+    encoder_hidden_states: Optional[torch.Tensor] = None,
+    encoder_attention_mask: Optional[torch.Tensor] = None,
+    use_cache: Optional[bool] = False,
+    output_attentions: Optional[bool] = False,
+    token_idx: Optional[bool] = False,
```
Comment (regisss) [inline optimum/habana/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py:22]:
```suggestion
    token_idx: Optional[torch.Tensor] = None,
```

File: optimum/habana/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+    self,
+    hidden_states: Optional[Tuple[torch.Tensor]],
+    layer_past: Optional[torch.Tensor] = None,
+    attention_mask: Optional[torch.Tensor] = None,
+    head_mask: Optional[torch.Tensor] = None,
+    encoder_hidden_states: Optional[torch.Tensor] = None,
+    encoder_attention_mask: Optional[torch.Tensor] = None,
+    use_cache: Optional[bool] = False,
+    output_attentions: Optional[bool] = False,
+    token_idx: Optional[bool] = False,
```
Comment (regisss) [inline optimum/habana/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py:95]:
```suggestion
    token_idx: Optional[torch.Tensor] = None,
```

File: optimum/habana/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+    position_ids: Optional[torch.Tensor] = None,
+    head_mask: Optional[torch.Tensor] = None,
+    inputs_embeds: Optional[torch.Tensor] = None,
+    encoder_hidden_states: Optional[torch.Tensor] = None,
+    encoder_attention_mask: Optional[torch.Tensor] = None,
+    use_cache: Optional[bool] = None,
+    output_attentions: Optional[bool] = None,
+    output_hidden_states: Optional[bool] = None,
+    return_dict: Optional[bool] = None,
+    token_idx: Optional[bool] = False,
```
Comment (regisss) [inline optimum/habana/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py:169]:
```suggestion
    token_idx: Optional[torch.Tensor] = None,
```

File: optimum/habana/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+        head_mask: Optional[torch.Tensor] = None,
+        inputs_embeds: Optional[torch.Tensor] = None,
+        encoder_hidden_states: Optional[torch.Tensor] = None,
+        encoder_attention_mask: Optional[torch.Tensor] = None,
+        labels: Optional[torch.Tensor] = None,
+        use_cache: Optional[bool] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+        token_idx: Optional[bool] = False,
```
Comment (regisss) [inline optimum/habana/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py:413]:
```suggestion
        token_idx: Optional[torch.Tensor] = None,
```

**Review by regisss — COMMENTED** [review]
@PhillipHoward Nice PR :fire: 

I just left a few minor comments that shouldn't take long to address.

Also, the code quality check failed so could you run the following from the root of the repo please?
```bash
pip install black ruff
make style
```

**Review by regisss — COMMENTED** [review]
@PhillipHoward Thanks for updating the PR!

I see the code style check failed again, could you run the following from the root of the repo please?
```bash
pip install black ruff
make style
```

**Comment by PhillipHoward** [comment]
@regisss sorry about that, I forgot to re-format the code after merging changes. It should pass now.

**Review by regisss — APPROVED** [review]
LGTM!

**Pull request merged**